File: backend/src/patrones_de_diseño/visitor/visitor.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""Demostración del patrón visitor en Python3"""

import random

class NoModificable: pass

# Las superclases de Flor no pueden ser modificadas:
class Flor(NoModificable):
    def aceptar(self, visitante): # Visitor (nombre en inglés del patrón)
        visitante.visitar(self)
    def polinizar(self, polinizador):
        print(self, "polinizado por", polinizador)
    def depredar(self, depredador):
        print(self, "depredado por", depredador)
    def __str__(self):
        return self.__class__.__name__

class Gladiolo(Flor): pass
class Tulipan(Flor): pass
class Crisantemo(Flor): pass

class Visitante:
    def __str__(self):
        return self.__class__.__name__

class Insecto(Visitante): pass
class Polinizador(Insecto): pass
class Depredador(Insecto): pass


class Abeja(Polinizador):
    def visitar(self, flor):
        flor.polinizar(self)

class Mosquito(Polinizador):
    def visitar(self, flor):
        flor.polinizar(self)

class Gusano(Depredador):
    def visitar(self, flor):
        flor.depredar(self)


def flowerGen(n):
    flores = Flor.__subclasses__()
    for i in range(n):
        yield random.choice(flores)()

abeja = Abeja()
mosquito = Mosquito()
gusano = Gusano()

for flor in flowerGen(10):
    flor.aceptar(abeja)
    flor.aceptar(mosquito)
    flor.aceptar(gusano)

File: backend/src/patrones_de_diseño/visitor/visitor.js
"use strict"

// Demostración del patrón visitor en Javascript

class NoModificable {};

// La superclase de flor no puede ser modificada
class Flor extends NoModificable {
    aceptar(visitante) {
        visitante.visitar(this);
    }
    polinizar(polinizador) {
        console.log(this, "polinizado por", polinizador)
    }
    depredar(depredador) {
        console.log(this, "depredado por", depredador)
    }
    inspect() {
    	return this.constructor.name;
    }
}
//Flor.prototype.inspect = function(){
//	return this.constructor.name;
//}

class Gladiolo extends Flor{}
class Tulipan extends Flor{}
class Crisantemo extends Flor{}

class Visitante {
    inspect() {
        return this.constructor.name;
    }
}

class Insecto extends Visitante {}
class Polinizador extends Insecto {}
class Depredador extends Insecto {}

class Abeja extends Polinizador {
    visitar(flor) {
        flor.polinizar(this);
    }
}
class Mosquito extends Polinizador {
    visitar(flor) {
        flor.polinizar(this);
    }
}
class Gusano extends Depredador {
    visitar(flor) {
        flor.depredar(this);
    }
}

function randomChoice(choices) {
    var index = Math.floor(Math.random() * choices.length);
    return choices[index];
}

function flowerGen(n) {
    var response = [];
    var flores = [Gladiolo, Tulipan, Crisantemo];
    for (var i=1; i<=n; i++){
        var flor = randomChoice(flores);
        response.push(new flor);
    }
    return response
}


var abeja = new Abeja();
var mosquito = new Mosquito();
var gusano = new Gusano();

var flores = flowerGen(10);
for (var i=0; i<flores.length; i++){
    flores[i].aceptar(abeja);
    flores[i].aceptar(mosquito);
    flores[i].aceptar(gusano);
}
